File: javascript/easy/minCostClimbingStairs.js
/**
 * @param {number[]} cost
 * @return {number}
 */
var minCostClimbingStairs = function(cost) {
    if(cost.length <= 0) return 0;
    const memo = new Array(cost.length), n = cost.length;
    for(let i = n-1; i >= 0; i--) {
        memo[i] = cost[i];
        let next = (i < n-1) ? memo[i+1] : 0;
        let next2 = (i < n-2) ? memo[i+2] : 0;
        memo[i] += Math.min(next, next2);
    }
    return Math.min(memo[0], memo[1]);
};

File: ruby/easy/min_cost_climbing_stairs.rb
# @param {Integer[]} cost
# @return {Integer}
def min_cost_climbing_stairs(cost)
    memo = []
    return [min_cost(0, cost, memo), min_cost(1, cost, memo)].min
end

def min_cost(i, cost, memo)
    if i == cost.length
        return 0
    elsif i > cost.length
        return (2 ** 31) - 1
    elsif !memo[i]
        memo[i] = cost[i] + [min_cost(i+1, cost, memo), min_cost(i+2, cost, memo)].min
    end
    return memo[i]
end
